### ISSUE TITLE
Update Root on ZFS Debian guide to follow the style of the Arch/Fedora/Red Hat guides

### ISSUE THREAD

**Issue by savyajha** [openzfs/openzfs-docs#264]
The Debian guide seems to be written in quite a different style compared to the RH/Fedora/Arch guides. Would it be acceptable if it was remade in the same way? The changes I would make if I were to do this would be:

- Remove separate guides for Buster and Bullseye. Consolidate them into one single guide. Keep the Stretch guide in its current form till oldoldstable's EOL date of 2022-06-30
- The dataset structure followed by the other guides would be followed
- `/etc/fstab` would be used for mounting all datasets
- Instructions to use both `initramfs-tools` and `dracut` would be added, since both are available in Debian

Does this sound like something worth pursuing? What other changes would be relevant for the existing Debian guide?

**Comment by ne9z** [comment]
Debian and Ubuntu guides are written by @rlaager, over [the course of last year](https://github.com/openzfs/openzfs-docs/pulls?q=is%3Apr+author%3Ane9z+) I improved upon them, to add support for Arch/Fedora/RHEL/NixOS. openSUSE guide is written by @Zaryob.

rlaager seems to be quite busy these days and there's relatively little activity as a result. We also have some difference in writing styles, as you can seen here: [1](https://github.com/openzfs/openzfs-docs/pull/104#issuecomment-759966648) and here: [2](https://github.com/openzfs/openzfs-docs/pull/262#issuecomment-1027379715), and in the guide themselves of course.

As the Ubuntu/Debian guides are the most visited pages so far (as shown by the number of issues filed for them at this issue tracker), I would suggest you to contact rlaager first and have a discussion about this rewrite.

Thanks for your contributions!

**Comment by savyajha** [comment]
Thanks for giving me something to work on! I shall await @rlaager's approval and once they say yay or nay I shall proceed. Just for context, I use zfs on Debian Bullseye daily on a semi-production system, and I used the Arch Linux how-to to set it up because it seemed cleaner to me than the current Debian write-up. I also have experience managing an Arch on zfs machine. I had used the Debian write-up for a Buster installation previously and found that the `/etc/fstab`-based mounting detailed in the Arch guide seemed more understandable and more logical as compared to the hybrid `zfs-mount-generator`/systemd service based version in the Debian write-up.

In addition, since the styles of the other distros seem similar to each other I thought it might help in bringing everything into the same style both in writing and the steps they recommend one follow.

**Comment by ne9z** [comment]
That's quite a lot of work to adapt Arch guide for Debian on-the-fly, I would say. Glad that you succeed at doing so!

> semi-production system

**Now time for some shameless self-promotion:**-) In that case I would highly recommend you to install [bieaz boot environment manager](https://gitlab.com/m_zhou/bieaz). This script, together with an appropriate [apt hook](https://unix.stackexchange.com/a/263747) and accompanying [commands](https://gitlab.com/m_zhou/rozb3-pac/-/blob/a9feb14c53ceebd5c8fa5d2008d1304eb3e3a800/libalpm/scripts/rozb3-pac#L40), would enable rollback of system updates at GRUB menu and preserve system configurations.

There are some prerequisites: first is to clone the [git repo](https://gitlab.com/m_zhou/bieaz) and install the scripts according to [makefile](https://gitlab.com/m_zhou/bieaz/-/blob/e28d2d63812bd3cc4504e42c784713bba98678a0/Makefile), a simple ``make install`` would do. (I haven't got the time to study how to do packaging for Debian yet.)

The second is to follow the GRUB installation steps laid out in arch guide, which are
```
# workaround
sed -i "s|rpool=.*|rpool=\`zdb -l \${GRUB_DEVICE} \| grep -E '[[:blank:]]name' \| cut -d\\\' -f 2\`|"  /etc/grub.d/10_linux

source /etc/os-release
# we will use ${ID}=debian in the file

mkdir -p /boot/efi/EFI/${ID}
mkdir -p /boot/grub

# legacy boot
for i in ${DISK}; do
 grub-install --boot-directory /boot/efi/EFI/${ID} --target=i386-pc $i
done

# efi boot
grub-install --boot-directory /boot/efi/EFI/${ID} --efi-directory /boot/efi/
grub-install --boot-directory /boot/efi/EFI/${ID} --efi-directory /boot/efi/ --removable

# generate grub menu compatible with boot environment manager
grub-mkconfig -o /boot/efi/EFI/${ID}/grub/grub.cfg
cp /boot/efi/EFI/${ID}/grub/grub.cfg /boot/grub/grub.cfg

# mirror efi partition content
ESP_MIRROR=$(mktemp -d)
cp -r /boot/efi/EFI $ESP_MIRROR
for i in /boot/efis/*; do
 cp -r $ESP_MIRROR/EFI $i
done
```

**Comment by rlaager** [comment]
TL;DR: No.


> Remove separate guides for Buster and Bullseye. Consolidate them into one single guide.

I only intend to support the latest stable version (e.g. Bullseye). Previously, when I cut a new HOWTO, I simply removed the old one. However, people requested that I leave the old ones for troubleshooting purposes. So these days, they get a warning that they're historical, and I don't (generally) apply changes to them.

Sometimes, as in the case between Buster and Bullseye, the changes are pretty small. So when looking at that particular pairing, it wouldn't really be that big of a deal to keep supporting the old version. But that's not true as a general rule. Sometimes the difference across distro versions is quite significant (e.g. when Ubuntu adopted zsys, lots of stuff changed).

> The dataset structure followed by the other guides would be followed /etc/fstab would be used for mounting all datasets

No, that's just wrong. On systemd systems, zfs-mount-generator should be used. systemd's fstab generator will get you close, but there are definitely good reasons to use zfs-mount-generator.

> Instructions to use both initramfs-tools and dracut would be added, since both are available in Debian

As far as I can see, initramfs-tools is the default in Debian for non-ZFS systems. Unless there is some significant ZFS-related advantage from switching to Dracut, I'd prefer not. For better or worse, it's a design principle of mine that the ZFS instructions should follow non-ZFS defaults as much as possible. In other words, the system should be as "normal" as possible, except for the fact that it uses ZFS. There are multiple reasons for this. First off, the more system the system is to a non-ZFS system, the less someone has to learn.

Second, the closer it is to a default system, the easier it would be to add support to the installer. Granted, I'm not sure if that'll ever happen on Debian, due to the licensing issues. And Ubuntu did add support (which AFAIK is still there), but it's unfortunately not a priority for them these days, since they never completed enhancements like encryption.
